.github/workflows/retry.yml

```diff
@@ -0,0 +1,184 @@
+# Please ping @DilumAluthge when making any changes to this file.
+
+# Here are some steps that we take in this workflow file for security reasons:
+# 1. We do not checkout any code.
+# 2. We do not run any external actions.
+# 3. We do not give the `GITHUB_TOKEN` any permissions.
+# 4. We only give the Buildkite API token (`BUILDKITE_API_TOKEN_RETRY`) the minimum necessary
+#    set of permissions.
+
+# Important note to Buildkite maintainers:
+# In order to make this work, you need to tell Buildkite that it should NOT create a brand-new
+# build when someone closes and reopens a pull request. To do so:
+# 1. Go to the relevant pipeline (e.g. https://buildkite.com/julialang/julia-master).
+# 2. Click on the "Pipeline Settings" button.
+# 3. In the left sidebar, under "Pipeline Settings", click on "GitHub".
+# 4. In the "GitHub Settings", under "Build Pull Requests", make sure that the "Skip pull
+#    request builds for existing commits" checkbox is checked. This is the setting that tells
+#    Buildkite that it should NOT create a brand-new build when someone closes and reopens a
+#    pull request.
+# 5. At the bottom of the page, click the "Save GitHub Settings" button.
+
+name: Retry Failed Buildkite Jobs
+
+on:
+  # When using the `pull_request_target` event, all PRs will get access to secret environment
+  # variables (such as the `BUILDKITE_API_TOKEN_RETRY` secret environment variable), even if
+  # the PR is from a fork. Therefore, for security reasons, we do not checkout any code in
+  # this workflow.
+  # TODO: change `pull_request` to `pull_request_target`.
```

Comment:
Can't someone just edit this file and open a PR with a modified `.yml` file?

Reply:
No. If someone (who does not have commit access) makes a malicious PR from a fork, then:
- The `GITHUB_TOKEN` will only have read-only access.
- No other secrets will be available. So, for example, the `BUILDKITE_API_TOKEN_RETRY` secret will not be available.

There are two types of events to consider here:
1. The `pull_request` event
2. The `pull_request_target` event

When someone makes a PR from a fork, the `pull_request` event will:
- Use the copy of the workflow file in their PR
- Only have a read-only `GITHUB_TOKEN`
- Not have access to any other secrets

So for the `pull_request` event, yes the workflow file has malicious content, but the `GITHUB_TOKEN` is read-only, and there is no access to any other secrets, so the only thing that the attacker can do is make read calls to the GitHub API.

When someone makes a PR from a fork, the `pull_request_target` event will:
- Use the copy of the workflow file in the `master` branch
- Have a `GITHUB_TOKEN` with write access
- Have access to secrets (including but not limited to the `BUILDKITE_API_TOKEN_RETRY` secret)

So, in the `pull_request_target` event, the job that runs has write access and access to secrets. However, the workflow file that is being used is the one from the `master` branch. So as long as we don't merge the attacker's PR, their malicious code will not be run.

Does that make sense?

Reply:
The main purpose of the `Therefore, for security reasons, we do not checkout any code in this workflow` comment is to remind me (and other committers) that we should not merge any PR that checks out code as part of the `pull_request_target` event.

Basically, it's a reminder to follow the security guidelines for the `pull_request_target` event discussed here: https://securitylab.github.com/research/github-actions-preventing-pwn-requests/

Reply:
Perfect!  I love the GH integration here.